### PR TITLE
v2tenancy: v2 namespace controller + tests

### PR DESCRIPTION
**Pull request by analogue**
### Changes proposed in this PR ###  
See acceptance criteria in https://hashicorp.atlassian.net/browse/NET-6633

There are a lot of unrelated changes in this PR to address test flakes due to 
1) the v1 ACL system not being bootstrapped
2) the v1 builtin default partition not existing

I've fixed both of these flakes in a general way (see additions to `TestServerWithMockConnMgrWatcher`) so that all existing tests will benefit and removed one-off workarounds in each individual test. I also refactored `TestServerClient` to have `ResourceClient` as a field that has the anti-flake measures applied. Existing tests were refactored to use this new v2 resource client.
 
### How I've tested this PR ###
- Unit tests for the v2 namespace controller that cover staying in sync with k8s namespaces
  - Separate build tagged enterprise tests that exercise non-default partitions.
  - Since non-default namespaces are now part of CE in v2, they're covered in non-ent tests.

- Manual test
```
1) SETUP: Install consul

`helm install rnconsul <local chart dir> --values values.min.yaml` where values are

# values.min.yaml
global:
  name: gnconsul
  image: consul-dev:latest
  imageK8S: consul-k8s-control-plane-dev:latest
  logLevel: "debug"
  experiments: ["resource-apis", "v2tenancy"]
  enableConsulNamespaces: true

server:
  logLevel: "debug"
  extraEnvironmentVars:
    CONSUL_LICENSE: <redacted>

ui:
  # not compatible with resource-apis
  enabled: false

2) SUCCESS: Create kube namespace and make sure it gets created in consul: 
  `kubectl create namespace test1`
  `consul resource list tenancy.v2beta1.Namespace` <-- shows `test1`

3) SUCCESS: Delete kube namespace and make sure it gets deleted in consul:
  `kubectl delete namespace test1`
  `consul resource list tenancy.v2beta1.Namespace` <-- `test1` absent
```  

### How I expect reviewers to test this PR ###
- Run the tests

### Reviewers ###
- Tagging @thisisnotashwin as the primary reviewer
- FYI @DanStough and @zalimeni for changes to unrelated tests to make sure I didn't regress anything wrt test flakes.
- CC @dhiaayachi for coverage on consul core team

### Checklist ###
- [x] Tests added
- [ ] [CHANGELOG entry added](https://github.com/hashicorp/consul-k8s/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

**Review by zalimeni — COMMENTED** [review]
A few minor questions but overall the test changes look good to me - thank you for all the cleanup!

File: control-plane/subcommand/inject-connect/command.go [hunk 1]
```diff
@@ -420,6 +423,10 @@ func (c *Command) validateFlags() error {
 		return errors.New("-consul-dataplane-image must be set")
 	}
 
+	if c.flagV2Tenancy && !c.flagResourceAPIs {
+		return errors.New("-enable-resource-apis must be set to 'true' if -enable-v2tenancy is set")
+	}
```
[the remark this answers is not in the document]
Comment (zalimeni) [reply control-plane/subcommand/inject-connect/command.go:428]: Looking at this and some recent info shared around core changes, I think I now see why this makes sense. Still wondering whether we need a separate flag in K8s - at least right now, given this check - but I follow the rationale for splitting in Core.

**Comment by analogue** [comment]
Update: Manually tested and added results to description.

File: control-plane/subcommand/inject-connect/command.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (analogue) [reply control-plane/subcommand/inject-connect/command.go:428]: Good question - this is a bit nuanced and a result of the split brain we introduced in Consul 1.17 where v2 catalog and mesh resource are backed by v1 tenancy and acls. 

This allowed multiport to ship in 1.17 in hybrid mode. With consul 1.18, we need to be explicit about whether we're in "hybrid" mode (v2 catalog + v2 mesh + v1 tenancy + v1 acls) or "full v2" mode (v2 catalog + v2 mesh + v2 tenancy + ACLs not supported).

As a result, all affected components need to be aware and flow accordingly:
- resources-apis=true => hybrid mode
- resource-apis=true and v2tenancy=true => full v2 mode and no ACLs

In short, this is how the `connect-inject` subcommand knows whether to spin up the v1 namespace controller or the v2 namespace controller. Never both.

File: control-plane/tenancy/namespace/namespace_controller.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+	// MirroringK8s causes Consul namespaces to be created to match the
+	// k8s namespace of any config entry custom resource. Config entries will
+	// be created in the matching Consul namespace.
+	MirroringK8s bool
+
+	// MirroringK8sPrefix is an optional prefix that can be added to the Consul
+	// namespaces created while mirroring. For example, if it is set to "k8s-",
+	// then the k8s `default` namespace will be mirrored in Consul's
+	// `k8s-default` namespace.
+	MirroringK8sPrefix string
```
[the remark this answers is not in the document]
Comment (analogue) [reply control-plane/tenancy/namespace/namespace_controller.go:53]: Didn't know `ConsulTenancyConfig` existed! I'll see if I can refactor this.

Separately, I got put off by the names being different from the source of truth (values.yaml) and explicitly changed the names to match. This would enable future maintainers to search for the verbatim setting name values.yaml" and quickly find where the it is actually tied to logic in the codebase. Regardless, I'll update things back to being inconsistently consistent 😄.

File: control-plane/subcommand/inject-connect/command.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (zalimeni) [reply control-plane/subcommand/inject-connect/command.go:428]: That makes sense, thanks for the extra context @analogue ! If that's the established pattern then it makes sense to keep it consistent here.

~ Small request: could we add a blurb comment along those lines here or somewhere else appropriate so that context is visible?

File: control-plane/tenancy/namespace/namespace_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (zalimeni) [reply control-plane/tenancy/namespace/namespace_controller.go:53]: > Separately, I got put off by the names being different from the source of truth (values.yaml) and explicitly changed the names to match. This would enable future maintainers to search for the verbatim setting name in values.yaml" and quickly find where it is actually tied to logic in the codebase. Regardless, I'll update things back to being inconsistently consistent 😄.

Ha, fair. FWIW I'd gladly approve a follow-up PR to rename for flag->struct consistency if you felt inclined to do it (no pressure though).

File: control-plane/helper/test/test_util.go [hunk 3]
```diff
@@ -20,24 +20,31 @@ import (
...
 type TestServerClient struct {
-	TestServer *testutil.TestServer
-	APIClient  *api.Client
-	Cfg        *consul.Config
-	Watcher    consul.ServerConnectionManager
+	TestServer     *testutil.TestServer
+	APIClient      *api.Client
+	Cfg            *consul.Config
+	Watcher        consul.ServerConnectionManager
+	ResourceClient pbresource.ResourceServiceClient
```
Comment (thisisnotashwin) [inline control-plane/helper/test/test_util.go:47]: ❤️

File: control-plane/subcommand/server-acl-init/command_ent_test.go [hunk 4]
```diff
@@ -1449,9 +1449,6 @@ func partitionedSetup(t *testing.T, bootToken string, partitionName string) *tes
 	server.Cfg.APIClientConfig.Token = bootToken
 	serverAPIClient, err := consul.NewClient(server.Cfg.APIClientConfig, 5*time.Second)
 	require.NoError(t, err)
-
 	_, _, err = serverAPIClient.Partitions().Create(context.Background(), &api.Partition{Name: partitionName}, &api.WriteOptions{})
-	require.NoError(t, err)
```
Comment (thisisnotashwin) [inline control-plane/subcommand/server-acl-init/command_ent_test.go:1454]: is there a reason we have dropped this require check?

File: control-plane/tenancy/namespace/namespace_controller.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+// SetupWithManager registers this controller with the manager.
+func (r *Controller) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Namespace{}).
+		Complete(r)
+}
+
+// consulNamespaceFor returns the Consul destination namespace for a provided Kubernetes namespace
+// depending on Consul Namespaces being enabled and the value of namespace mirroring.
+func (r *Controller) consulNamespaceFor(kubeNamespace string) string {
```
Comment (thisisnotashwin) [inline control-plane/tenancy/namespace/namespace_controller.go:124]: didnt we want just 1:1 mapping for Consul and Kubernetes namespace in v2?

Comment (thisisnotashwin) [reply control-plane/tenancy/namespace/namespace_controller.go:124]: as in Kubernetes Namespace name == Consul Namespace name?

File: control-plane/tenancy/namespace/namespace_controller.go (hunk 2 above)
Comment (thisisnotashwin) [inline control-plane/tenancy/namespace/namespace_controller.go:53]: I could be wrong, but i thought in v2 tenancy we are going to map the Kubernetes namespace to a Consul namespace with the same name and not support destination namespaces and prefixes? as in, always be mirroring?

**Review by thisisnotashwin — COMMENTED** [review]
This looks excellent Semir!! I just have some questions about supporting use-cases that arent "mirrored".

File: control-plane/subcommand/inject-connect/command.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (analogue) [reply control-plane/subcommand/inject-connect/command.go:445]: Updated in https://github.com/hashicorp/consul-k8s/pull/3435/commits/fa47ad03964628c984ded0f5e69d5afb8f37393c

[the remark this answers is not in the document]
Comment (zalimeni) [reply control-plane/subcommand/inject-connect/command.go:445]: Thank you! 🎨

File: control-plane/tenancy/namespace/namespace_controller.go (hunk 5 above)
Comment (analogue) [reply control-plane/tenancy/namespace/namespace_controller.go]: Removed since mapping is now straight 1:1

File: control-plane/subcommand/server-acl-init/command_ent_test.go (hunk 4 above)
Comment (analogue) [reply control-plane/subcommand/server-acl-init/command_ent_test.go:1454]: Good catch! Didn't intend to to that. Restored in https://github.com/hashicorp/consul-k8s/pull/3435/commits/a31041315728b81eae2e00a0ed361edf514c1edd

**Review by thisisnotashwin — APPROVED** [review]
excellent!!!
